packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {IMoxieBondingCurve} from "./interfaces/IMoxieBondingCurve.sol";
+import {SecurityModule} from "./SecurityModule.sol";
+import {ITokenManager} from "./interfaces/ITokenManager.sol";
+import {IERC20Extended} from "./interfaces/IERC20Extended.sol";
+import {IStaking} from "./interfaces/IStaking.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+/**
+ * @title Staking
+ * @author Moxie Team
+ * @notice Staking contract allows staking of subject tokens for a lock period.
+ */
+contract Staking is IStaking, SecurityModule, ReentrancyGuard {
+    using SafeERC20 for IERC20Extended;
+
+    bytes32 public constant CHANGE_LOCK_DURATION =
+        keccak256("CHANGE_LOCK_DURATION");
+
+    ITokenManager public tokenManager;
+    IMoxieBondingCurve public moxieBondingCurve;
+    IERC20Extended public moxieToken;
+
+    uint256 public lockCount;
+
+    mapping(uint256 lockId => LockInfo lockInfo) public locks;
+    mapping(uint256 lockPeriodInSec => bool allowed) public lockPeriodsInSec;
```

Comment:
I think we should have a mapping that gives the lockId belong to a user / subject. Otherwise, how it will be difficult to retrieve. also we need to have getter function for this as well.

Reply:
Subgraph will have that info. Its an indexing layer problem